crates/scion-proto/src/scmp/messages.rs

```diff
@@ -83,57 +152,125 @@ macro_rules! call_method_on_scmp_variants {
+            Self::ExternalInterfaceDown(x) => x.$name($($param),*),
+            Self::InternalConnectivityDown(x) => x.$name($($param),*),
+            Self::Unknown(x) => x.$name($($param),*),
+        }
+    };
+}
+
+macro_rules! call_method_on_scmp_info_variants {
+    ($self:ident.$name:ident($($param:ident),*)) => {
+        match $self {
+            Self::EchoRequest(x) => x.$name($($param),*),
+            Self::EchoReply(x) => x.$name($($param),*),
+            Self::TracerouteRequest(x) => x.$name($($param),*),
+            Self::TracerouteReply(x) => x.$name($($param),*),
+        }
+    };
+}
+
 macro_rules! lift_fn_from_scmp_variants {
     (
+        $call_method:ident;
         $(#[$outer:meta])*
         $vis:vis fn $name:ident(self$(: $self_ty:ty)? $(,$param:ident : $param_type:ty)*) -> $return_type:ty
     ) => {
         $(#[$outer])*
         $vis fn $name(self$(: $self_ty)? $(,$param : $param_type)*) -> $return_type {
-            call_method_on_scmp_variants!(self.$name($($param),*))
+            $call_method!(self.$name($($param),*))
         }
-    };
+    }
 }
 
-impl ScmpMessage {
-    lift_fn_from_scmp_variants!(
-        /// Returns the type of the corresponding message.
-        pub fn get_type(self: &Self) -> ScmpType
-    );
+macro_rules! implement_methods_for_scmp_enum {
+    (
+        $name:ident, $call_method:ident
+    ) => {
+        impl ScmpMessageBase for $name {
+            lift_fn_from_scmp_variants!(
+                $call_method;
+                fn get_type(self: &Self) -> ScmpType
+            );
+
+            lift_fn_from_scmp_variants!(
+                $call_method;
+                fn code(self: &Self) -> u8
+            );
+        }
 
-    lift_fn_from_scmp_variants!(
-        /// Returns the code field of the corresponding message.
-        pub fn code(self: &Self) -> u8
-    );
+        impl $name {
+            lift_fn_from_scmp_variants!($call_method;
 
-    lift_fn_from_scmp_variants!(
-        /// Returns true if the checksum successfully verifies, otherwise false.
-        pub fn verify_checksum(self: &Self, address_header: &AddressHeader) -> bool
-    );
+                /// Returns true if the checksum successfully verifies, otherwise false.
+                pub fn verify_checksum(self: &Self, address_header: &AddressHeader) -> bool
+            );
 
-    lift_fn_from_scmp_variants!(
-        /// Returns true if the checksum successfully verifies, otherwise false.
-        pub fn set_checksum(self: &mut Self, address_header: &AddressHeader) -> ()
-    );
+            lift_fn_from_scmp_variants!($call_method;
 
-    /// Returns true iff `self` is an error message.
-    pub fn is_error(&self) -> bool {
-        self.get_type().is_error()
-    }
+                /// Returns true if the checksum successfully verifies, otherwise false.
+                pub fn set_checksum(self: &mut Self, address_header: &AddressHeader) -> ()
+            );
+        }
```

Comment:
Remark: I think the docstring is not correct.

Reply:
Thanks. Fixed by actually implementing `MessageChecksum` for `$name`.